### PR TITLE
fix(custom_header): not correctly parsing parameters

### LINKED ISSUE #19: bug(scan): Global Variable - custom_header does not parse correctly
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Submitted by @Zeetaz

When trying to run a scan utilzing the global variable "custom_header" it will be parsed differently depending on which tool it is assigned to.

For example, running the following YAML scanner config:
```yaml
enable_http_crawl: true           # All tools
threads: 10                       # All tools

custom_header: 'User-Agent: Hello_I_Will_Die_PoC'

fetch_url: {
  'uses_tools': ['katana'],
}
```
It will break:

```bash
celery_1       |   warnings.warn(
celery_1       |
celery_1       | fetch_url                          | WARNING | Task fetch_url is RUNNING
celery_1       | fetch_url                          | ERROR | 'str' object has no attribute 'items'
celery_1       | Traceback (most recent call last):
celery_1       |   File "/usr/src/app/reNgine/celery_custom_task.py", line 129, in __call__
celery_1       |     self.result = self.run(*args, **kwargs)
celery_1       |   File "/usr/src/app/reNgine/tasks.py", line 1812, in fetch_url
celery_1       |     f'{key}: {value}' for key, value in custom_header.items()
celery_1       | AttributeError: 'str' object has no attribute 'items'
celery_1       | fetch_url                          | WARNING | Task fetch_url status is FAILED | Error: AttributeError("'str' object has no attribute 'items'")
```

If you instead were to run a vulnerability scan instead for **fetch_url** it will function correctly for nuclei.

It is possible to make this work as a temporary solution:

```yaml
enable_http_crawl: true           # All tools
threads: 10                       # All tools

custom_header: {
  '"User-Agent': 'Hello_Working_PoC"'
}

fetch_url: {
  'uses_tools': ['katana'],
}
```
But this reuqires you to run two separate scans for vulnerability scans and fetching URL's... I am also not sure if it will break other tools or if it is only katana that is currently broken an expects "items" instead of strings.

### Expected Behavior

The expected behavior, for most tools and for katan, would be to run the configuration as:
 ```yaml
custom_header = 'User-Agent: testing'
etc, etc.
 ```
And it should be appended to the tool (katan in this case) as 
 ```bash
-H "User-Agent: testing"
 ```

### Steps To Reproduce

1. Run the following scanner configuration on any target:

 ```yaml
enable_http_crawl: true           # All tools
threads: 10                       # All tools

custom_header: 'User-Agent: Hello_I_Will_Die_PoC'

fetch_url: {
  'uses_tools': ['katana'],
}
 ```
2. Note that the scan fails as it does not expect a str value.

3. Run the following - modified - configuration:

```yaml
enable_http_crawl: true           # All tools
threads: 10                       # All tools

custom_header: {
  '"User-Agent': 'Hello_Working_PoC"'
}

fetch_url: {
  'uses_tools': ['katana'],
}
```

4. Note that this will work, however it does not follow the expected syntax. This is also an extremely bad work-around as it will not work with nuclei or any other tools as they won't parse it the same.

(it will append the header to katana as `-H "User-Agent: Hello_Working_PoC"`)

### Environment

```markdown
- reNgine: 2.0.3
- OS: 
- Python: 
- Docker Engine: 
- Docker Compose: 
- Browser:
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by psyray**
Fixes #19 
When you set custom header in the scan, value does not been parsed correctly, making httpx crash and not resolving http request.
As httpx is the base of the scan, no other scan will be launched because no URL was found.

I've reworked the run_command and stream_command to be more rock solid.
I've also removed duplicated code

Should fix other problems when parameters contains spaces in their value.

Now custom headers is fully effective and working fine

- [x] Test command with global headers
- [x] Test command with task headers
- [x] Test command with both headers
- [x] Test command without headers

## Summary by Sourcery

Fix the issue with custom header parsing that caused httpx crashes, refactor command execution functions for improved stability, and enhance logging for better debugging.

Bug Fixes:
- Fix parsing of custom headers to prevent crashes in httpx and ensure proper HTTP request resolution.

Enhancements:
- Improve the robustness of the run_command and stream_command functions by refactoring and removing duplicated code.
- Add logging for debugging purposes in various functions, including fetch_url and generate_header_param.

### CODE REVIEW

**Comment by AnonymousWP** [comment]
Due to ocervell's experience, I think it's better if he (or another developer) from the team reviews this.

**Comment by psyray** [comment]
@sourcery-ai review

File: web/reNgine/tasks.py [hunk 1]
```diff
@@ -4173,139 +4180,91 @@ def remove_duplicate_endpoints(
 
 
 @app.task(name='run_command', bind=False, queue='run_command_queue')
```
[the remark this answers is not in the document]
Comment (psyray) [reply web/reNgine/tasks.py:4186]: I will refactor this later

File: web/reNgine/common_func.py [hunk 2]
```diff
@@ -1081,6 +1083,9 @@
     Returns:
         str: Command-line parameter for the specified tool.
     """
+    logger.debug(f"Generating header parameters for tool: {tool_name}")
```
[the remark this answers is not in the document]
Comment (psyray) [reply web/reNgine/common_func.py:1086]: I will do this later

File: web/reNgine/common_func.py [hunk 3]
```diff
@@ -1167,3 +1176,92 @@
...
+    Args:
+        command (str or list): The command to execute.
+        shell (bool): Whether to use shell execution.
+        cwd (str): The working directory for the command.
+
+    Returns:
+        subprocess.Popen: The Popen object for the executed command.
+    """
+    return subprocess.Popen(
+        command,
```
[the remark this answers is not in the document]
Comment (psyray) [reply web/reNgine/common_func.py:1261]: You could test and merge
I will resolve this when I will do refactoring

**Review by AnonymousWP — CHANGES_REQUESTED** [review]
I think this is the only thing I can find. Not sure how much effort I should put into testing every possibility.

File: web/reNgine/tasks.py
[the remark this answers is not in the document]
Comment (psyray) [reply web/reNgine/tasks.py:1]: Fixed$
There's a check that trigger the error, when parsing headers, in make logs but now double quotes are sanitized while saving/updating an engine

**Review by AnonymousWP — APPROVED** [review]

**Pull request merged**